### PR TITLE
Framework for accelerating compression using QAT

### PR DESCRIPTION
**Pull request by hhb584520**
Accelerate compression to provide improved efficiency, scalability and
performance. Intel QuickAssist Technology (Intel QAT) offloading
Compression from CPU which can save cycles, time and space.

As the complexity of applications continues to grow, systems need more
and more computational resources for workloads, including cryptography
and data compression.

The drivers and patches offered here assist application developers to
take advantage of Intel® QuickAssist Integrated Accelerator that is
offered with platforms based on Intel® processors.

To Learn how to use Intel® QuickAssist Technology and run example code,
review our tutorial videos on Intel Developer Zone – Intel® QuickAssist
Technology Technical Getting Started Tutorials -
https://software.intel.com/en-us/networking/quickassist

Signed-off-by: Haibin Huang <haibin.huang@intel.com>

### CODE REVIEW

**Comment by BigVan** [comment]
I see...
Actually, I think  zfile_compress_software() is a special case in  zfile_compress_qat() which the number of batches is only one..
So how about just changing the parameters of ICompressor::compress() as compress(const unsigned char *const raw_data[], size_t steplist[],  unsigned char *compressed_data[], size_t compressed_len[], ssize_t cur, size_t dst_len) instead of a new interface?

**Comment by BigVan** [comment]
> 

and check if QAT enable in compress() method itself

File: src/overlaybd/fs/zfile/compressor.cpp [hunk 1]
```diff
@@ -76,17 +76,165 @@ class Compressor_lz4 : public ICompressor {
     }
 };
 
+#ifdef ENABLE_QAT
```
Comment (BigVan) [inline src/overlaybd/fs/zfile/compressor.cpp:79]: add a member 'qat_enable' in Compressor_lz4 and move 'bool check_qat()' in this file?

**Comment by hhb584520** [comment]
> I see... Actually, I think zfile_compress_software() is a special case in zfile_compress_qat() which the number of batches is only one.. So how about just changing the parameters of ICompressor::compress() as compress(const unsigned char *const raw_data[], size_t steplist[], unsigned char *compressed_data[], size_t compressed_len[], ssize_t cur, size_t dst_len) instead of a new interface?

good idea, Yes, this is the easiest way, but you need to modify all the places where you call compress. I wonder if you can accept it

File: src/overlaybd/fs/zfile/compressor.cpp (hunk 1 above)
Comment (hhb584520) [reply src/overlaybd/fs/zfile/compressor.cpp:79]: ok, good point

**Comment by BigVan** [comment]
We only use the method 'ICompressor::compress()'  in zfile.cpp, so I think modifying the original parameters of 'compress()' is OK. Or maybe I can try to change some code in zfile.cpp so you only need to implement QAT in ICompressor?

**Comment by hhb584520** [comment]
> We only use the method 'ICompressor::compress()' in zfile.cpp, so I think modifying the original parameters of 'compress()' is OK. Or maybe I can try to change some code in zfile.cpp so you only need to implement QAT in ICompressor?

It would be best if you could modify zFile to be compatible with batch commit processing, and it would be easier for us to modify.

**Comment by hhb584520** [comment]
> > 
> 
> and check if QAT enable in compress() method itself

ok

**Comment by BigVan** [comment]
I create a PR in your repo.
BTW, please join our slack channel in CNCF
Login to the workspace through https://slack.cncf.io/  and search 'overlaybd'.
Thanks

**Comment by hhb584520** [comment]
> I create a PR in your repo. BTW, please join our slack channel in CNCF Login to the workspace through https://slack.cncf.io/ and search 'overlaybd'. Thanks

ok, thank you, I have seen the PR.

**Comment by hhb584520** [comment]
@BigVan according to your comments and patch, we have modified it, please review it again.

**Comment by BigVan** [comment]
LGTM

**Pull request merged**